fixtures/provider.py

```diff
@@ -1,94 +1,237 @@
+    pf = ProviderFilter(classes=[prov_class])
+    return setup_one_or_skip(request, filters=[pf], use_global_filters=use_global_filters)
+
+
+def _generate_provider_fixtures():
+    """ Generate provider setup and clear fixtures based on what classes are available
+
+    This will make fixtures like "cloud_provider" and "has_no_cloud_provider" available to tests.
+    """
+    for prov_type, prov_class in get_provider_type_class_map().iteritems():
+        def gen_setup_provider(prov_class):
+            @pytest.fixture(scope='module')
+            def _setup_provider(request):
+                """ Sets up one of the matching providers """
+                return setup_one_by_class_or_skip(request, prov_class)
+            return _setup_provider
+        fn_name = '{}_provider'.format(prov_type)
+        globals()[fn_name] = gen_setup_provider(prov_class)
+
+        def gen_has_no_providers(prov_class):
+            @pytest.fixture
+            def _has_no_providers():
+                """ Clears all providers of given class from the appliance """
+                BaseProvider.clear_providers_by_class(prov_class, validate=True)
+            return _has_no_providers
+        fn_name = 'has_no_{}_provider'.format(prov_type)
+        globals()[fn_name] = gen_has_no_providers(prov_class)
+
+
+# Let's generate all the provider setup and clear fixtures within the scope of this module
+_generate_provider_fixtures()
+
+
+# When we want to setup a provider provided by testgen
+# ----------------------------------------------------
 @pytest.fixture(scope='function')
 def setup_provider(request, provider):
     """Function-scoped fixture to set up a provider"""
-    _setup_provider(provider.key, request)
+    return setup_or_skip(request, provider)
 
 
 @pytest.fixture(scope='module')
 def setup_provider_modscope(request, provider):
-    """Function-scoped fixture to set up a provider"""
-    _setup_provider(provider.key, request)
+    """Module-scoped fixture to set up a provider"""
+    return setup_or_skip(request, provider)
 
 
 @pytest.fixture(scope='class')
 def setup_provider_clsscope(request, provider):
     """Module-scoped fixture to set up a provider"""
-    _setup_provider(provider.key, request)
+    return setup_or_skip(request, provider)
 
 
 @pytest.fixture
 def setup_provider_funcscope(request, provider):
-    """Function-scoped fixture to set up a provider
-
-    Note:
-
-        While there are cases where this is useful, provider fixtures should
-        be module-scoped the majority of the time.
-
-    """
-    _setup_provider(provider.key, request)
+    """Function-scoped fixture to set up a provider"""
+    return setup_or_skip(request, provider)
+# -----------------------------------------------
 
 
 @pytest.fixture(scope="session")
-def any_provider_session():
-    BaseProvider.clear_providers()  # To make it clean
-    providers.setup_a_provider(validate=True, check_existing=True)
+def any_provider_session(request):
```

Comment:
**REQUIRED:** I think perhaps we should rename this to only_one_provider_session